lib/sim.c

```diff
@@ -0,0 +1,969 @@
+        return -1;
+    }
+    int64_t balance, limit;
+    int trustline_res = get_trustlines(a, c, e, &balance, &limit);
+    if (trustline_res < 0)
+        return trustline_res;
+
+    struct Trustline *tl = (struct Trustline *)malloc(sizeof(struct Trustline));
+    memcpy(tl->high_account, (uint32_t *)a, b);
+    memcpy(tl->low_account, (uint32_t *)c, d);
+    memcpy(tl->amount.iou.issuer, (uint32_t *)c, d);
+    tl->amount.is_xrp = 0;
+    memcpy(tl->amount.iou.currency, (uint32_t *)e, f);
+    tl->amount.iou.amount = balance;
+    tl->limit = limit;
+
+    const int64_t addr = (int64_t)tl;
+    // Write trustline struct address to the write_ptr so it can be retrieved in the slot_set for future use.
+    // Memory locations will be cleaned in the FREE_TRANSACTION_OBJ macro.
+    INT64_TO_BUF(write_ptr, addr);
+    memset((uint32_t *)(write_ptr + 8), 0, write_len - 8);
+
+    return write_len;
+}
+
+int64_t state_set(uint32_t read_ptr, uint32_t read_len, uint32_t kread_ptr, uint32_t kread_len)
+{
+    // Acording to the xrpl hook specs,
+    // If kread_len is greater than max key len or read len is greater than max data len, return TOO_BIG.
+    // If kread_len is 0, return TOO_SMALL.
+    if (kread_len > STATE_KEY_LEN || read_len > MAX_STATE_VAL_LEN)
+        return TOO_BIG;
+    else if (kread_len == 0)
+        return TOO_SMALL;
+
+    // Send the state set request.
+    const int len = 1 + STATE_KEY_LEN + read_len;
+    uint8_t buf[len];
+    // Populate the type header.
+    buf[0] = STATE_SET;
+    // Populate the state data.
+    memcpy(&buf[1], (uint32_t *)kread_ptr, kread_len);
+    // If kread_len is less than STATE_KEY_LEN, populate o's to the rest.
+    if (kread_len < STATE_KEY_LEN)
+        memset(&buf[1 + kread_len], 0, STATE_KEY_LEN - kread_len);
+    // Populate the data to rest.
+    memcpy(&buf[1 + STATE_KEY_LEN], (uint32_t *)read_ptr, read_len);
+    write_stdout(buf, len);
+
+    // Read the response from STDIN.
+    int8_t ret;
+    const int data_len = read_stdin(&ret, 1);
+
+    // Return the error code according to the return code.
+    if (ret == RET_OVERFLOW)
+        return TOO_BIG;
+    else if (ret == RET_UNDERFLOW)
+        return TOO_SMALL;
+    else if (ret < -1)
+        return -1;
+
+    return 0;
+}
+
+int64_t state(uint32_t write_ptr, uint32_t write_len, uint32_t kread_ptr, uint32_t kread_len)
+{
+    // Acording to the xrpl hook specs,
+    // If kread_len is greater than max key len, return TOO_BIG.
+    if (kread_len > STATE_KEY_LEN)
+        return TOO_BIG;
+
+    // Send the state set request.
+    const int len = 1 + kread_len;
+    uint8_t buf[len];
+    // Populate the type header.
+    buf[0] = STATE_GET;
+    // Populate the state key.
+    memcpy(&buf[1], (uint32_t *)kread_ptr, kread_len);
+    write_stdout(buf, len);
+
+    // Read the response from STDIN.
+    uint8_t data_buf[MAX_STATE_VAL_LEN];
```

Comment:
Forgot this! this should be `uint8_t data_buf[MAX_STATE_VAL_LEN + 1];` (space for return code)